contributors/design-proposals/live-and-inplace-vertical-scaling.md

```diff
@@ -0,0 +1,325 @@
+So we need the means to express and implement the best approach for resizing for each workload and resource.
+
+# Use Cases
+
+* After deploying a service/application through `StatefulSet`, I want to resize, by giving a updated pod spec with new resource sizes, the resources (for now, CPU and memory) allocated to each pod of the statefulset without restarting the pods.
+
+* A (vertical) pod auto-scaler (c.f. [a related issue](https://github.com/kubernetes/features/issues/21)) can utilize this feature, while currently it does resize a pod with restart only.
+
+# Objectives
+
+1. Enable live and in-place resource resizing on a pod.
+2. Add support for live and in-place resource resizing in `StatefulSet` controller.
+
+# API and Usage
+
+To express the policy for resizing in a pod spec, we introduce resource attribute `resizePolicy` with the following choices for value:
+* RestartOnly. (the current behavior, default)
+* LiveResizeable.
+
+This attribute will be available per resource (such as cpu, memory) and so is adequate to indicate whether the workload can handle and prefer a change in each resource’s allocation for it without restarting.
+With potentially multiple containers and multiple resizeable resources for each in a Pod, the response to an update of the pod spec will be determined by the a precedence order among the attribute values with RestartOnly dominating LiveResizeable, i.e., if two resources have been resized in the update to the spec and one of them has a policy of RestartOnly then the pod would be restarted to realize both updates.
+
+We can optionally introduce an action annotation `resizeAction` with following choices for value:
+* Restart. (default)
+* LiveResize.
+* LiveResizePreferred.
+
+If used, this would be included as part of the patch or appropriate update command providing the spec update for the resize.
+It would indicate the preference of user at the time of resize.
+Specifically, Restart for `resizeAction` would indicate the pod be restarted for the corresponding resizing of resource(s), LiveResize would indicate the pod not be restarted the resize be realized live, and LiveResizePreferred would indicate that the resize be realized preferrably live but if that fails for any reason to accomplish it with a restart.
+
+An example of the usage of resizePolicy attribute in a pod spec:
+
+```yaml
+resources:
+    requests:
+        cpu: 100m
+        memory: 1Gi
+    limits:
+        cpu: 1000m
+        memory: 1Gi
+    resizePolicy:
+        cpu: LiveResizeable
+        memory: RestartOnly
+```
+
+For the above example, if there is a change to cpu request or limit it can be vertically scaled only if the memory request and limit remained the same, otherwise the RestartOnly policy for memory would override the policy for CPU, and the Pod (container, if container-alone restart is allowed) would need to be restarted.
```

Comment:
I am inclined to deny a resize request that changed the pods QoS class.  Noting here also concerns about desired G pod characteristics for NUMA, exclusive cores (if on a node with static cpu manager), hugepages, and devices.  It seems for some of those options RestartOnly is the only viable option.

Reply:
Container storage is another area

Reply:
for exclusive cores, why RestartOnly should be the only viable option?

Reply:
what if I update pod spec to change its QoS class from `BestEffort` to `Guaranteed`?

Reply:
As mentioned below at 'Related issues' section, QoS class change by live-and-in-place resizing is not supported because in order to change QoS class of a pod, the parent directory of the cgroup directory of the pod also need to change, which is not possible.

Reply:
For NUMA, CPU manager, and other stuffs which were introduced after 1.7, still we're working to make sure that our vertical scaling feature works well with them. Perhaps, could you provide some insights on problems that might come ??